### PR TITLE
AP-3698

### PR DESCRIPTION
**Pull request by chiichang**
This update involves:
AP-3698: incorrect activities after filtering event attributes
AP-3721: Case variant IDs did not get updated after filtering event attributes
AP-3700: Support the attribute filter shortcut of PD

### CODE REVIEW

File: Apromore-Custom-Plugins/Process-Discoverer-Portal-Plugin/src/main/resources/processdiscoverer/zul/components/logStats.zul [hunk 1]
```diff
@@ -80,7 +80,7 @@
       <vlayout spacing="0">
         <n:div class="ap-pd-logstats-header ap-link" title="Filter by Events">
           <button id="btnEventHeading" sclass="ap-icon ap-icon-flag" ></button>
-          <label id="lblEventHeading" sclass="title" value="Events"/>
+          <label id="lblEventHeading" sclass="title" value="Activity inst."/>
```
Comment (brucenguyen1) [inline Apromore-Custom-Plugins/Process-Discoverer-Portal-Plugin/src/main/resources/processdiscoverer/zul/components/logStats.zul:83]: From testing with logs having both start and complete events, it shows the total events in the log, not 'activity instance', or is it "Event" now called "Activity instance"?

image

File: Apromore-Custom-Plugins/Process-Discoverer-Portal-Plugin/src/main/java/org/apromore/plugin/portal/processdiscoverer/impl/apmlog/LogFilterControllerWithAPMLog.java [hunk 2]
```diff
@@ -248,6 +261,23 @@ private RuleValue findRuleValueByOpeType(OperationType operationType, Set<RuleVa
...
+
+    private static String convertAttributeKey(String label) {
+        switch (label) {
+            case "Resources": return "org:resource";
+            case "Activities": return "concept:name";
+            case "Groups": return "org:group";
+            default: return label;
+        }
+    }
+
```
Comment (brucenguyen1) [inline Apromore-Custom-Plugins/Process-Discoverer-Portal-Plugin/src/main/java/org/apromore/plugin/portal/processdiscoverer/impl/apmlog/LogFilterControllerWithAPMLog.java:280]: Test with "Production_Data" log, open it in PD, select "Qty completed" or "Qty rejected" in the perspective list, note that the Activity heading in the stat box changes to "Qty completed", click on it, error is shown below.

File: Apromore-Custom-Plugins/Process-Discoverer-Portal-Plugin/src/main/resources/processdiscoverer/zul/components/logStats.zul (hunk 1 above)
Comment (brucenguyen1) [reply Apromore-Custom-Plugins/Process-Discoverer-Portal-Plugin/src/main/resources/processdiscoverer/zul/components/logStats.zul:83]: Confirmed with Marcello:

image

**Comment by brucenguyen1** [comment]
NullPointerException in the Filter window: for example, open any log in the Dashboard, open the filter window (next to the log name at the top), set a filter criterion that removes everything (e.g. remove all case IDs, remove all events with all Activities), select OK -> NPE.

File: Apromore-Custom-Plugins/Process-Discoverer-Portal-Plugin/src/main/resources/processdiscoverer/zul/components/logStats.zul (hunk 1 above)
Comment (brucenguyen1) [reply Apromore-Custom-Plugins/Process-Discoverer-Portal-Plugin/src/main/resources/processdiscoverer/zul/components/logStats.zul:83]: Inconsistent UI: both "Event" and "Activity Instance" are used

image

File: Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/filter/rules/desc/AttributeArcDurationDesc.java [hunk 3]
```diff
@@ -50,7 +50,7 @@ public static String getDescription(LogFilterRule logFilterRule) {
         sb.append(logFilterRule.getChoice() == Choice.RETAIN ? "Retain " : "Remove ");
         sb.append(" all cases where the direct-follows relation '");
         sb.append(fromVal + "' -> '" + toVal + "'");
```
Comment (franklma) [inline Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/filter/rules/desc/AttributeArcDurationDesc.java]: Such calls may profitably be turned into chained append calls on the existing StringBuffer/Builder/Appendable, saving the cost of an extra StringBuffer/Builder allocation.

**Comment by brucenguyen1** [comment]
Error in Filter window: open a simple log in PD, add a filter criterion to select 1 case only, then add another filter criterion to select 1 event only (as shown below). Then on the "Temporal Statistics" box in PD, click on "Log timeframe" => Error below or "Case duration" => endless loop.

File: Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/filter/PTrace.java [hunk 4]
```diff
@@ -54,664 +57,414 @@
...
-
-            this.activityNameIndexList = previousActiivtyNameIndexList;
-        } else {
-            reset();
-        }
+    public DoubleArrayList getWaitingTimes() {
+        List<AActivity> validActs = StatsUtil.getValidActivitiesOf(this);
+        double[] waitTimesArray = validActs.stream()
+                .filter(x -> StatsUtil.getValidPreviousActivity(x, this) != null)
+                .mapToDouble(x -> StatsUtil.getArcDurationOf(StatsUtil.getValidPreviousActivity(x, this), x))
```
Comment (franklma) [inline Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/filter/PTrace.java]: Argument 'StatsUtil.getValidPreviousActivity(x, this)' might be null

File: Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/filter/PTrace.java [hunk 5]
```diff
@@ -54,664 +57,414 @@
...
+    @Override
     public UnifiedSet<String> getEventNameSet() {
-        return this.eventNameSet;
+        return null;
     }
 
     public UnifiedMap<String, String> getAttributeMap() {
-        Map<String, String> collect = attributeMap.entrySet().stream()
+        Map<String, String> collect = aTrace.getAttributeMap().entrySet().stream()
                 .filter(x -> !x.getKey().equals("concept:name") && !x.equals("case:variant") )
```
Comment (franklma) [inline Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/filter/PTrace.java:298]: 'equals()' between objects of inconvertible types 'Entry<String, String>' and 'String'

File: Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/immutable/ImmutableTrace.java [hunk 6]
```diff
@@ -321,6 +322,18 @@ public long getCaseIdDigit() {
         return null;
     }
 
+    public String getActivityNameIndexString(HashBiMap<String, Integer> nameIndexBiMap) {
```
Comment (franklma) [inline Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/immutable/ImmutableTrace.java:325]: line 187, In method getAttributeMap()
'equals()' between objects of inconvertible types 'Entry<String, String>' and 'String'

File: Apromore-Extras/APMLogModule/src/test/java/org/apromore/apmlog/AttributeArcDurationTest.java [hunk 7]
```diff
@@ -254,14 +258,19 @@ public static void testRetain3(APMLog apmLog, APMLogUnitTest parent) throws Unsu
...
         pLog.updateStats();
-        DurSubGraph subGraph = pLog.getAttributeGraph().getNextValueDurations("concept:name", "A", pLog);
-        DoubleArrayList durList = subGraph.getValDurListMap().get("A");
 
-        assertEquals(1000 * 60 * 60 * 2d, durList.average(), 0 /* comparison tolerance */);
+        String key = "concept:name";
+        UnifiedSet<EventAttributeValue> eavSet = pLog.getEventAttributeValues().get(key);
+        Map<String, List<CustomTriple>> data = StatsUtil.getTargetNodeDataBySourceNode(key, "A", pLog, eavSet);
+
+        double[] durArray = data.get("A").stream().mapToDouble(CustomTriple::getDuration).toArray();
```
Comment (franklma) [inline Apromore-Extras/APMLogModule/src/test/java/org/apromore/apmlog/AttributeArcDurationTest.java:271]: Method invocation 'get' may produce 'NullPointerException'

**Comment by brucenguyen1** [comment]
NullPointerException in Filter window: 
Open a simple log in PD, set a filter criterion to retain only one single case (e.g. right click on an box), then click on the "Activities" heading in the Log stats, the Filter window is opened with the activity above selected and Cases option = Remove, note that other activities are also shown, select other activities, click OK -> NPE dialog shows up.

File: Apromore-Custom-Plugins/Process-Discoverer-Portal-Plugin/src/main/java/org/apromore/plugin/portal/processdiscoverer/impl/apmlog/LogFilterControllerWithAPMLog.java [hunk 8]
```diff
@@ -91,19 +91,41 @@ private void onInvokeEvent(String payload) {
...
+
+        Map<String, Object> payload = getEditorOptionPayload(filterType, param);
+        if (payload == null || payload.isEmpty()) {
+            return;
+        }
+
+        Clients.showBusy("Launch Filter Dialog ...");
+
+        LogFilterRule rule = getLastMatchedRuleWithValues(filterType, payload,
```
Comment (brucenguyen1) [inline Apromore-Custom-Plugins/Process-Discoverer-Portal-Plugin/src/main/java/org/apromore/plugin/portal/processdiscoverer/impl/apmlog/LogFilterControllerWithAPMLog.java:103]: When click on the "Activities" heading in the Log Stats box in PD, should it show only the "last matched rule" for "Activities" or should it show all the matched rules for "Activities". The latter seems more reasonable.

Comment (brucenguyen1) [reply Apromore-Custom-Plugins/Process-Discoverer-Portal-Plugin/src/main/java/org/apromore/plugin/portal/processdiscoverer/impl/apmlog/LogFilterControllerWithAPMLog.java:103]: Ok for now.

**Review by brucenguyen1 — APPROVED** [review]

**Review by franklma — APPROVED** [review]

**Pull request merged**